### ISSUE TITLE
Server crash with SASL using SCRAM-SHA-1

### ISSUE THREAD

**Issue by logicwonder** [processone/ejabberd#3794]
## Environment

- ejabberd version: 20.07
- Erlang version: 9.2
- OS: Linux (Debian)
- Installed from: source 

## Errors from crash.log

2022-03-23 00:00:01 =ERROR REPORT====
** Generic server <0.26776.3938> terminating
** Last message in was {'$gen_event',{xmlstreamelement,{xmlel,<<"auth">>,[{<<"xmlns">>,<<"urn:ietf:params:xml:ns:xmpp-sasl">>},{<<"mechanism">>,<<"SCRAM-SHA-1">>}],[{xmlcdata,<<"biwsbj1hMDI0OGZiZjc3NzBkNzVmLHI9NSQpWCo2PjJoZDxmPjxtbXt2U3lEc0hVI0stWGYiJ1s=">>}]}}}
** When Server state == #{mgmt_max_queue => 5000,mgmt_ack_timeout => 60000,stream_id => <<"6921299824212369485">>,lserver => <<"xyz.zzzz.com">>,stream_encrypted => true,mgmt_stanzas_out => 0,mgmt_timeout => 300000,pres_a => {0,nil},socket => {socket_state,fast_tls,{tlssock,#Port<0.114953215>,#Ref<0.3445500336.2590375940.113136>},65536,#Ref<0.3445500336.2265841669.39807>,{state,32000,32000,32000,1647973801079144},{{{0,0,0,0,0,65535,2807,35528},443},{{0,0,0,0,0,65535,35751,48266},42586}}},user => <<>>,zlib => false,tls_enabled => false,mgmt_state => inactive,ip => {{0,0,0,0,0,65535,35751,48266},42586},xmlns => <<"jabber:client">>,tls_verify => false,mgmt_max_timeout => 300000,mgmt_queue_type => ram,server => <<"xyz.zzzz.com">>,socket_monitor => #Ref<0.3445500336.2265710597.39808>,stream_state => wait_for_sasl_request,stream_compressed => false,stream_restarted => true,stream_timeout => {40000,-572912487109},mgmt_stanzas_in => 0,owner => <0.26776.3938>,mgmt_stanzas_req => 0,csi_queue => {0,#{}},stream_version => {1,0},tls_required => true,stream_authenticated => false,lang => <<"en-GB">>,stream_header_sent => true,access => c2s,tls_options => [compression_none],mgmt_resend => true,codec_options => [ignore_els],resource => <<>>,stream_direction => in,mod => ejabberd_c2s,shaper => c2s_shaper,csi_state => active}
** Reason for termination ==
** {{badmatch,false},[{base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},{xmpp_sasl_scram,mech_step,2,[{file,"src/xmpp_sasl_scram.erl"},{line,99}]},{xmpp_sasl,server_step,2,[{file,"src/xmpp_sasl.erl"},{line,120}]},{xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,893}]},{xmpp_stream_in,handle_info,2,[{file,"src/xmpp_stream_in.erl"},{line,404}]},{p1_server,handle_msg,8,[{file,"src/p1_server.erl"},{line,696}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
2022-03-23 00:00:01 =CRASH REPORT====
  crasher:
    initial call: xmpp_stream_in:init/1
    pid: <0.26776.3938>
    registered_name: []
    exception exit: {{{badmatch,false},[{base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},{xmpp_sasl_scram,mech_step,2,[{file,"src/xmpp_sasl_scram.erl"},{line,99}]},{xmpp_sasl,server_step,2,[{file,"src/xmpp_sasl.erl"},{line,120}]},{xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,893}]},{xmpp_stream_in,handle_info,2,[{file,"src/xmpp_stream_in.erl"},{line,404}]},{p1_server,handle_msg,8,[{file,"src/p1_server.erl"},{line,696}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]},[{p1_server,terminate,7,[{file,"src/p1_server.erl"},{line,878}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [ejabberd_c2s_sup,ejabberd_sup,<0.10921.0>]
    message_queue_len: 0
    messages: []
    links: [<0.19335.0>]
    dictionary: [{already_terminated,true},{rand_seed,{#{bits => 58,jump => #Fun<rand.8.15449617>,next => #Fun<rand.5.15449617>,type => exrop,uniform => #Fun<rand.6.15449617>,uniform_n => #Fun<rand.7.15449617>,weak_low_bits => 1},[46712597558555109|49803856023054843]}},{'$internal_queue_len',0}]
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 4999
  neighbours:
2022-03-23 00:00:01 =SUPERVISOR REPORT====
     Supervisor: {local,ejabberd_c2s_sup}
     Context:    child_terminated
     Reason:     {{badmatch,false},[{base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},{xmpp_sasl_scram,mech_step,2,[{file,"src/xmpp_sasl_scram.erl"},{line,99}]},{xmpp_sasl,server_step,2,[{file,"src/xmpp_sasl.erl"},{line,120}]},{xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,893}]},{xmpp_stream_in,handle_info,2,[{file,"src/xmpp_stream_in.erl"},{line,404}]},{p1_server,handle_msg,8,[{file,"src/p1_server.erl"},{line,696}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
     Offender:   [{pid,<0.26776.3938>},{id,undefined},{mfargs,{ejabberd_c2s,start_link,undefined}},{restart_type,temporary},{shutdown,5000},{child_type,worker}]

## Bug description

We are continuously getting this crash in all eJabberd instances in a 3 node cluster. Recently we have kicked multiple user accounts by setting an empty password in SQL DB. Not sure whether this is related.
Could anyone guide on the issue.

**Comment by prefiks** [comment]
Hello,

This appears to be problem with data in db (we aren't able to properly decode what was stored), somehow data got corrupted here.

Could you tell me how you did set that empty password in db? Maybe something happened here. I think this crash was generated for user a0248fbf7770d75f

**Comment by logicwonder** [comment]
Any idea which field was not properly decoded. 

The password for certain users were set to empty by updating the users table in SQL DB. Later this approach was changed and ban_account API was used instead. 

The user a0248fbf7770d75f is an active user and does not fall under empty password/banned category. From the logs we can see the user was authenticated successfully (Accepted C2S Scram) two hours before crash.

**Comment by logicwonder** [comment]
Using a query on users table since we are using a SQL DB for eJabberd. Can I set a random password using ban_account API for all such accounts to resolve this issue?

**Comment by logicwonder** [comment]
Thank you @badlop for your detailed reply. 

> Does it affect specific accounts? Maybe weird data stored in database, or the used auth (username, password, resource or other  detail)

After random sampling of requests, it seems the issue mostly affects records that had empty passwords. I shall set random password (ban_account) to all such accounts first before looking for a pattern.

> Does it affect specific XMPP clients? Maybe problem in the client, or the used auth

We are currently using Smack and XMPP Framework in our mobile clients. Will study this.

> In summary: what do those crash login attempts have in common between them, and differently to non-crashy logins?
Will try to find this too by using the tracing strategy proposed.

Also how would eJabberd behave if the banned accounts are continuing with failed authentication attempts?

**Comment by badlop** [comment]
> Can I set a random password using ban_account API for all such accounts to resolve this issue?

You can use ban_account, the random password will look like:
```
BANNED_ACCOUNT--20220324T10:40:28--414161245595741019--ABC
```

I looked at that password in the SQL database, used it in the XMPP client, and of course it's able to login

> Also how would eJabberd behave if the banned accounts are continuing with failed authentication attempts?

I don't understand the question. If the client doesn't provide the correct password, it rejects the login and reports that in the log file.

Now thinking, maybe there's some option that affects this: did you enable scram? did you enable auth cache? Once you investigate this, if you get no clue, show your ejabberd configuration.

**Comment by logicwonder** [comment]
In my current configuration, use_cache is enabled globally and auth_use_cache is not configured. 

Auth configuration is given below:
auth_method: sql
auth_password_format: scram

In the crash it shows error seems to be related to scram.The crash still keeps coming. I have traced a crash authentication request for a user and found that all the previous authentication attempts by the user were successful. The correlated crash.log, error.log and ejabberd.log are given below for 0.2144.4070

crash.log
------------
2022-03-24 17:14:26 =ERROR REPORT====
** Generic server <0.2144.4070> terminating
** Last message in was {'$gen_event',{xmlstreamelement,{xmlel,<<"auth">>,[{<<"xmlns">>,<<"urn:ietf:params:xml:ns:xmpp-sasl">>},{<<"mechanism">>,<<"**SCRAM-SHA-1"**>>}],[{xmlcdata,<<"biwsbj1hMDk1Y2JjNTBhYzNiY2IzLHI9TSlrVjpyYVg2fW9cRHlNcmVGOzIrXEM+JEU8XmtUOHM=">>}]}}}
** When Server state == #{mgmt_max_queue => 5000,mgmt_ack_timeout => 60000,stream_id => <<"2238380551087925121">>,lserver => <<"xxxx.xxxx.xxx">>,stream_encrypted => true,mgmt_stanzas_out => 0,mgmt_timeout => 300000,pres_a => {0,nil},socket => {socket_state,fast_tls,{tlssock,#Port<0.120820487>,#Ref<0.3445500336.1436418055.113219>},65536,#Ref<0.3445500336.3354525697.169929>,{state,32000,32000,32000,1648122266607447},{{{0,0,0,0,0,65535,2807,35528},443},{{0,0,0,0,0,65535,40231,2657},47706}}},user => <<>>,zlib => false,tls_enabled => false,mgmt_state => inactive,ip => {{0,0,0,0,0,65535,40231,2657},47706},xmlns => <<"jabber:client">>,tls_verify => false,mgmt_max_timeout => 300000,mgmt_queue_type => ram,server => <<"xxxx.xxxx.xxx">>,socket_monitor => #Ref<0.3445500336.3354394625.169930>,stream_state => wait_for_sasl_request,stream_compressed => false,stream_restarted => true,stream_timeout => {40000,-572764021314},mgmt_stanzas_in => 0,owner => <0.2144.4070>,mgmt_stanzas_req => 0,csi_queue => {0,#{}},stream_version => {1,0},tls_required => true,stream_authenticated => false,lang => <<"en-IN">>,stream_header_sent => true,access => c2s,tls_options => [compression_none],mgmt_resend => true,codec_options => [ignore_els],resource => <<>>,stream_direction => in,mod => ejabberd_c2s,shaper => c2s_shaper,csi_state => active}
** Reason for termination ==
** {{badmatch,false},[{base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},{xmpp_sasl_scram,mech_step,2,[{file,"src/xmpp_sasl_scram.erl"},{line,99}]},{xmpp_sasl,server_step,2,[{file,"src/xmpp_sasl.erl"},{line,120}]},{xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,893}]},{xmpp_stream_in,handle_info,2,[{file,"src/xmpp_stream_in.erl"},{line,404}]},{p1_server,handle_msg,8,[{file,"src/p1_server.erl"},{line,696}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
2022-03-24 17:14:26 =CRASH REPORT====
  crasher:
    initial call: xmpp_stream_in:init/1
    pid: <0.2144.4070>
    registered_name: []
    exception exit: {{{badmatch,false},[{base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},{xmpp_sasl_scram,mech_step,2,[{file,"src/xmpp_sasl_scram.erl"},{line,99}]},{xmpp_sasl,server_step,2,[{file,"src/xmpp_sasl.erl"},{line,120}]},{xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,893}]},{xmpp_stream_in,handle_info,2,[{file,"src/xmpp_stream_in.erl"},{line,404}]},{p1_server,handle_msg,8,[{file,"src/p1_server.erl"},{line,696}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]},[{p1_server,terminate,7,[{file,"src/p1_server.erl"},{line,878}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [ejabberd_c2s_sup,ejabberd_sup,<0.10921.0>]
    message_queue_len: 0
    messages: []
    links: [<0.19335.0>]
    dictionary: [{already_terminated,true},{rand_seed,{#{bits => 58,jump => #Fun<rand.8.15449617>,next => #Fun<rand.5.15449617>,type => exrop,uniform => #Fun<rand.6.15449617>,uniform_n => #Fun<rand.7.15449617>,weak_low_bits => 1},[8053949769913404|150083206699646680]}},{'$internal_queue_len',0}]
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 4999
  neighbours:
2022-03-24 17:14:26 =SUPERVISOR REPORT====
     Supervisor: {local,ejabberd_c2s_sup}
     Context:    child_terminated
     Reason:     {{badmatch,false},[{base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},{xmpp_sasl_scram,mech_step,2,[{file,"src/xmpp_sasl_scram.erl"},{line,99}]},{xmpp_sasl,server_step,2,[{file,"src/xmpp_sasl.erl"},{line,120}]},{xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,893}]},{xmpp_stream_in,handle_info,2,[{file,"src/xmpp_stream_in.erl"},{line,404}]},{p1_server,handle_msg,8,[{file,"src/p1_server.erl"},{line,696}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
     Offender:   [{pid,<0.2144.4070>},{id,undefined},{mfargs,{ejabberd_c2s,start_link,undefined}},{restart_type,temporary},{shutdown,5000},{child_type,worker}]


error.log
-----------
2022-03-24 17:14:26.607 [error] <0.2144.4070>@base64:decode_binary:212 gen_server <0.2144.4070> terminated with reason: no match of right hand value false in base64:decode_binary/2 line 212
2022-03-24 17:14:26.607 [error] <0.2144.4070>@base64:decode_binary:212 CRASH REPORT Process <0.2144.4070> with 0 neighbours exited with reason: no match of right hand value false in base64:decode_binary/2 line 212 in p1_server:terminate/7 line 878
2022-03-24 17:14:26.607 [error] <0.19335.0>@base64:decode_binary:212 Supervisor ejabberd_c2s_sup had child undefined started with {ejabberd_c2s,start_link,undefined} at <0.2144.4070> exit with reason no match of right hand value false in base64:decode_binary/2 line 212 in context child_terminated


ejabberd.log
-----------------
`2022-03-24 17:14:26.044 [info] <0.13073.3330>@ejabberd_listener:accept:256 (<0.2144.4070>) Accepted proxied connection [::ffff:xxx.xxx.xxx.xxx]:47706 -> [::ffff:xxx.xxx.xxx.xxx]:5222`

**Comment by badlop** [comment]
My last attempt to reproduce the problem failed. I've compiled ejabberd 20.07, configured it to use MySQL for auth and enabled scram. I'll use Tkabber as client:

- Register an account and logins correctly.
- Then I set password = '' in the users MySQL table: client login is rejected, no crash
- Then I set serverkey and salt also to '' in users MySQL table: client login is rejected, no crash
- Finally I run ban_account, which sets a password, but keeps empty serverkey and salt: client login is rejected, no crash

So let's go back to the error message:

```
** Reason for termination ==
** {
 {badmatch,false},[
  {base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},
  {xmpp_sasl_scram,mech_step,2,[{file,"src/xmpp_sasl_scram.erl"},{line,99}]},
  {xmpp_sasl,server_step,2,[{file,"src/xmpp_sasl.erl"},{line,120}]},
  {xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,893}]},
  ...
```

The problem surfaces in xmpp_sasl_scram:mech_step/2, line 99, when it calls base64:decode_binary/2. ejabberd 20.07 depends on xmpp 1.4.9, let's look what does that file and lines say:
```erlang
					      {base64:decode(Pass#scram.storedkey),
```

https://github.com/processone/xmpp/blob/0029f3e5021e7152373187888473c9a2e5ca07a4/src/xmpp_sasl_scram.erl#L99
